File: training-data/python/command_pattern.py
from abc import ABC, abstractmethod
from typing import List


class Command(ABC):
    @abstractmethod
    def execute(self): ...

    @abstractmethod
    def undo(self): ...


class Light:
    def __init__(self):
        self.on = False


class TurnOnCommand(Command):
    def __init__(self, light: Light):
        self.light = light

    def execute(self):
        self.light.on = True

    def undo(self):
        self.light.on = False


class CommandHistory:
    def __init__(self):
        self.history: List[Command] = []

    def execute(self, command: Command):
        command.execute()
        self.history.append(command)

    def undo_last(self):
        if self.history:
            self.history.pop().undo()


if __name__ == "__main__":
    light = Light()
    history = CommandHistory()
    history.execute(TurnOnCommand(light))
    print(light.on)
    history.undo_last()
    print(light.on)

File: training-data/rust/command_pattern.rs
trait Command {
    fn execute(&self, light: &mut Light);
    fn undo(&self, light: &mut Light);
}

struct Light {
    on: bool,
}

struct TurnOnCommand;

impl Command for TurnOnCommand {
    fn execute(&self, light: &mut Light) {
        light.on = true;
    }
    fn undo(&self, light: &mut Light) {
        light.on = false;
    }
}

struct CommandHistory {
    history: Vec<Box<dyn Command>>,
}

impl CommandHistory {
    fn new() -> Self {
        CommandHistory { history: vec![] }
    }

    fn execute(&mut self, command: Box<dyn Command>, light: &mut Light) {
        command.execute(light);
        self.history.push(command);
    }

    fn undo_last(&mut self, light: &mut Light) {
        if let Some(command) = self.history.pop() {
            command.undo(light);
        }
    }
}

fn main() {
    let mut light = Light { on: false };
    let mut history = CommandHistory::new();
    history.execute(Box::new(TurnOnCommand), &mut light);
    println!("{}", light.on);
    history.undo_last(&mut light);
    println!("{}", light.on);
}
